microsoft_teams_importer: Convert users.

PieterCK opened the pull request:
This adds a command to convert Microsoft Teams export data. This version only converts:
- User accounts
- User roles

Next: #36316
**Screenshots and screen captures:**

| MS Teams | Zulip |
|--------|--------|
| <img width="1095" height="481" alt="image" src="https://github.com/user-attachments/assets/fe3437e2-ed46-4e3c-b453-a12ef64827b5" />| <img width="1063" height="696" alt="image" src="https://github.com/user-attachments/assets/a1426fbd-8fd2-466a-9e14-0d5f26103c5a" /> | 

The `help/import-microsoft-teams-data` page is not yet listed in the help center's sidebar.
<details>

<summary>
Importer documentation screenshot
</summary>
It's a long screenshot, open it in a new tab to see the full resolution.
<img width="1041" height="9747" alt="image" src="https://github.com/user-attachments/assets/995e6c50-4b82-427f-aa42-5a5151acaffd" />

</details>
- The "Import to Zulip, Self-hosting" instructions:
  <img width="1035" height="1645" alt="image" src="https://github.com/user-attachments/assets/131fec0f-9829-42a7-8264-308e3abfeb0d" />



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

mateuszmandera commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -0,0 +1,241 @@
+            email=microsoft_teams_user_email,
+            full_name=user_full_name,
+            id=zulip_user_id,
+            is_active=True,
```
do Microsoft Teams exports not have some concept of a deactivated user?

mateuszmandera commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -0,0 +1,241 @@
+            user_role = UserProfile.ROLE_MEMBER
+
+        microsoft_teams_user_email: str | None = (
+            user["Mail"] if user["Mail"] else user["UserPrincipalName"]
```
is `UserPrincipalName` guaranteed to be an email address?

mateuszmandera commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -0,0 +1,241 @@
+    odata_parameter = [
+        ODataQueryParameter(parameter="$filter", expression=f"userType eq '{member_type}'"),
+    ]
+    result = get_microsoft_graph_api_data(
```
Can you link to API documentation for the endpoints we're using? I guess the particular thing I'm wondering is whether we need pagination when there's a large amount of users? (Missing pagination was a sneaky bug we had in the Slack importer)

mateuszmandera reviewed (commented):
I posted some quick comments. Looks like a solid approach overall to me!

PieterCK commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -0,0 +1,241 @@
+@dataclass
+class ODataQueryParameter:
+    parameter: Literal["$filter", "$search", "$select"]
+    expression: str
```
```diff
 @dataclass
 class ODataQueryParameter:
+    # This is used to compute a request's OData query. It specifies the
+    # amount, type, and order of the data returned for the resource
+    # identified by the URL.
+    # https://learn.microsoft.com/en-us/graph/query-parameters?tabs=http
     parameter: Literal["$filter", "$search", "$select"]
     expression: str
```

PieterCK commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -0,0 +1,241 @@
+            id=zulip_user_id,
+            is_active=True,
+            role=user_role,
+            is_mirror_dummy=True,
```
Changed to `is_mirror_dummy=False` and added a comment in `convert_users`:

```diff
 def convert_users(
     microsoft_teams_user_role_data: MicrosoftTeamsUserRoleData,
     realm: dict[str, Any],
@@ -122,6 +145,11 @@ def convert_users(
     timestamp: int,
     users_list: list[MicrosoftTeamsFieldsT],
 ) -> MicrosoftTeamsUserIdToZulipUserIdT:
+    """
+    For now, this function only expects active user accounts because it's used to process
+    users listed in the `users/usersList.json` export file. That file includes only active
+    users -- no bot and no deleted user accounts.
+    """
     zerver_user_profile: list[ZerverFieldsT] = []
@@ -157,7 +181,7 @@ def convert_users(
             id=zulip_user_id,
             is_active=True,
             role=user_role,
-            is_mirror_dummy=True,
+            is_mirror_dummy=False,
             realm_id=realm_id,
             short_name=user_full_name,
             timezone="UTC",

```

PieterCK commented:
Thanks for reviewing this! I think I've addressed all of your comments, this is ready for review again.

Manually tested again:
<img width="959" height="578" alt="image" src="https://github.com/user-attachments/assets/e61fcf66-74b3-4572-a816-c527cf57771b" />

PieterCK commented:
@mateuszmandera -- Just pinging to check if you’ve had a chance to do another pass on this again.

mateuszmandera commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -0,0 +1,267 @@
+        microsoft_teams_user_id_to_zulip_user_id[microsoft_teams_user_id] = zulip_user_id
+
+        if user_role == UserProfile.ROLE_REALM_OWNER:
+            has_admin = True
```
let's call this `has_owner` since this is about `ROLE_REALM_OWNER` not `_ADMIN`.  And the `logging.info("Converted realm has no administrators!")` can be adjusted accordingly (`no owners!`)

PieterCK replied:
```diff
@@ -164,7 +167,7 @@ def convert_users(
     zerver_user_profile: list[ZerverFieldsT] = []
     microsoft_teams_user_id_to_zulip_user_id: MicrosoftTeamsUserIdToZulipUserIdT = defaultdict(int)
     found_emails: dict[str, int] = {}
-    has_admin = False
+    has_owner = False
 
     logging.info("######### IMPORTING USERS STARTED #########\n")
     for user in users_list:
@@ -206,7 +209,7 @@ def convert_users(
         microsoft_teams_user_id_to_zulip_user_id[microsoft_teams_user_id] = zulip_user_id
 
         if user_role == UserProfile.ROLE_REALM_OWNER:
-            has_admin = True
+            has_owner = True
 
         recipient_id = NEXT_ID("recipient")
         subscription_id = NEXT_ID("subscription")
@@ -222,8 +225,8 @@ def convert_users(
             microsoft_teams_user_email,
         )
 
-    if not has_admin:
-        logging.info("Converted realm has no administrators!")
+    if not has_owner:
+        logging.info("Converted realm has no owners!")
 
     validate_user_emails_for_import(list(found_emails))
     realm["zerver_userprofile"] = zerver_user_profile
@@ -258,7 +277,7 @@ class MicrosoftTeamsImporterUnitTest(ZulipTestCase):
                 microsoft_teams_user_role_data=microsoft_teams_user_role_data
             )
diff --git a/zerver/tests/test_microsoft_teams_importer.py b/zerver/tests/test_microsoft_teams_importer.py
index bda76cf7ab..e1280eaa45 100644
--- a/zerver/tests/test_microsoft_teams_importer.py
+++ b/zerver/tests/test_microsoft_teams_importer.py
         self.assertIn(
-            "INFO:root:Converted realm has no administrators!",
+            "INFO:root:Converted realm has no owners!",
             info_logs.output,
         )
```

mateuszmandera commented on zerver/tests/test_microsoft_teams_importer.py:
```diff
@@ -0,0 +1,328 @@
+        # For now we only convert active users listed in the usersList.json file in the export.
+        # So there shouldn't be any imported accounts with is_mirror_dummy=True.
+        non_user_account = self.get_imported_realm_user_field_values("id", is_mirror_dummy=True)
+        self.assertListEqual(non_user_account, [])
```
I think we want an assert like for `is_active=False`, since we expect all accounts to be active, right?

(also `mirror_dummy_accounts` will be a better name than `non_user_account`)

PieterCK replied:
Added `is_active=False` to the filter and renamed to `mirror_dummy_accounts`:

```diff
-        # For now we only convert active users listed in the usersList.json file in the export.
-        # So there shouldn't be any imported accounts with is_mirror_dummy=True.
-        non_user_account = self.get_imported_realm_user_field_values("id", is_mirror_dummy=True)
-        self.assertListEqual(non_user_account, [])
+        # For now the importer doesn't generate any mirror dummy accounts.
+        mirror_dummy_accounts = self.get_imported_realm_user_field_values(
+            "id", is_mirror_dummy=True, is_active=False
+        )
+        self.assertListEqual(mirror_dummy_accounts, [])
```

PieterCK replied:
> I think we want an assert like for is_active=False, since we expect all accounts to be active, right?

Tweaked this test to check all accounts are active:
```diff
@@ -150,13 +164,18 @@ class MicrosoftTeamsImporterIntegrationTest(ZulipTestCase):
 
     def test_imported_users(self) -> None:
         self.do_import_realm_fixture()
-        imported_user_emails = set(self.get_imported_realm_user_field_values("email"))
+        imported_user_emails = set(
+            self.get_imported_realm_user_field_values(
+                "email", is_mirror_dummy=False, is_active=True
+            )
+        )
         self.assertSetEqual(imported_user_emails, set(EXPORTED_MICROSOFT_TEAMS_USER_EMAIL.values()))
```

PieterCK commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -70,7 +70,18 @@ def get_microsoft_graph_api_data(
+    # https://learn.microsoft.com/en-us/graph/paging?tabs=http#server-side-paging
+    next_link: str | None = response_data.get("@odata.nextLink")
+    if next_link is not None:
+        result.extend(get_microsoft_graph_api_data(next_link, token=token))
```
```diff
diff --git a/zerver/data_import/microsoft_teams.py b/zerver/data_import/microsoft_teams.py
index ce12a50c5e..a6a2a4e0b3 100644
--- a/zerver/data_import/microsoft_teams.py
+++ b/zerver/data_import/microsoft_teams.py
@@ -56,71 +56,74 @@ def get_microsoft_graph_api_data(
     if not kwargs.get("token"):
         raise AssertionError("Microsoft authorization token missing in kwargs")
     token = kwargs.pop("token")
-    result = []
+    accumulated_result = []
     parameters = {}
     if odata_parameters is not None:
         for parameter in odata_parameters:
             assert parameter.parameter not in parameters
             parameters[parameter.parameter] = parameter.expression
 
-    response = requests.get(
-        api_url, headers={"Authorization": f"Bearer {token}"}, params=parameters
-    )
-
-    if response.status_code != requests.codes.ok:
-        logging.info("HTTP error: %s, Response: %s", response.status_code, response.text)
-        raise Exception("HTTP error accessing the Microsoft Graph API.")
-
-    response_data = response.json()
-    result.extend(response_data["value"])
-
     # If a request is paginated, "@odata.nextLink" will be included in the response,
     # it points to the next page of result. Even if the `$top` query is not specified,
     # depending on the endpoint and the result size, it may be paged by the server.
     # https://learn.microsoft.com/en-us/graph/paging?tabs=http#server-side-paging
-    next_link: str | None = response_data.get("@odata.nextLink")
-    if next_link is not None:
-        result.extend(get_microsoft_graph_api_data(next_link, token=token))
-    return result
+    next_link: str | None = api_url
+    while next_link is not None:
+        response = requests.get(
+            next_link, headers={"Authorization": f"Bearer {token}"}, params=parameters
+        )
+
+        if response.status_code != requests.codes.ok:
+            logging.info("HTTP error: %s, Response: %s", response.status_code, response.text)
+            raise Exception("HTTP error accessing the Microsoft Graph API.")
+
+        response_data = response.json()
+        accumulated_result.extend(response_data["value"])
+        next_link = response_data.get("@odata.nextLink")
+
+        # Don't apply initial request's parameters to subsequent paginated requests.
+        if next_link is not None:
+            parameters = {}
+
+    return accumulated_result
```
